### PR TITLE
Functional: validate None inputs against InputLayer(optional=True) and improve error handling

### PR DESCRIPTION
**Pull request by MalyalaKarthik66**
- Implements the TODO in `Functional._convert_inputs_to_tensors` to validate `None` inputs using the `InputLayer(optional=...)` flag.

- Previously, `None` values were passed through without validation and could lead to downstream errors (e.g., AttributeError during structure mismatch handling). This change ensures `None` is only allowed for optional inputs, while required inputs raise a `ValueError`. Existing behavior is preserved for tensors not originating from `InputLayer`.

- Also improves structure mismatch message safety for `None` and ensures user-facing `ValueError` is not masked during symbolic build.

- Added tests covering required inputs receiving `None`, mixed optional/required structures, and the `fit()` path. Verified across all backends.

### CODE REVIEW

File: keras/src/models/functional_test.py [hunk 1]
```diff
@@ -687,6 +687,62 @@ def test_disconnected_output(self):
...
+            def compute_output_shape(self, x_shape):
+                return x_shape
+
+        i1 = Input((2,), name="input1")
+        i2 = Input((2,), name="input2", optional=False)
+        outputs = OptionalInputLayer()(i1, i2)
+        model = Model({"input1": i1, "input2": i2}, outputs)
+
+        # Passing None for a required input must raise
+        with pytest.raises(ValueError):
```
Comment (hertschuh) [inline keras/src/models/functional_test.py]: Please change to `self.assertRaisesRegex` with a short part of the message. The truth is that without your change, it would already raise, but with a message that was not very clear.

File: keras/src/models/functional_test.py [hunk 2]
```diff
@@ -687,6 +687,62 @@ def test_disconnected_output(self):
...
+        # Small trainable model where input is required. fit should error
+        x = layers.Input((3,), name="x")
+        y = layers.Dense(1)(x)
+        model = Model(x, y)
+        model.compile(optimizer="sgd", loss="mse")
+
+        y_data = np.ones((4, 1))
+
+        # Passing None for required input should raise during fit
+        with pytest.raises(ValueError):
```
Comment (hertschuh) [inline keras/src/models/functional_test.py]: Use `self.assertRaisesRegex` with a short part of the message.

File: keras/src/models/functional.py [hunk 3]
```diff
@@ -225,12 +225,26 @@ def _maybe_warn_inputs_struct_mismatch(self, inputs, raise_exception=False):
                 tree.lists_to_tuples(self._inputs_struct),
             )
         except:
+
+            def _safe_name(x):
+                if x is None:
+                    return "None"
+                if isinstance(x, backend.KerasTensor):
+                    return x.name
+                return repr(x)
```
Comment (hertschuh) [inline keras/src/models/functional.py]: I don't think you need 3 cases here, it's either `None` or has a name.

So you can do:
```python
            model_inputs_struct = tree.map_structure(
                lambda x: "None" if x is None else x.name, self._inputs_struct
            )
```

File: keras/src/models/functional.py [hunk 4]
```diff
@@ -225,12 +225,26 @@ def _maybe_warn_inputs_struct_mismatch(self, inputs, raise_exception=False):
...
+
+            def _safe_inputs_repr(x):
+                if x is None:
+                    return "None"
+                if isinstance(x, backend.KerasTensor):
+                    return f"Tensor(shape={x.shape})"
+                return repr(x)
```
Comment (hertschuh) [inline keras/src/models/functional.py]: Same here, this should work:

```python
            inputs_struct = tree.map_structure(
                lambda x: "None" if x is None else f"Tensor(shape={x.shape})", inputs
            )
```

File: keras/src/models/functional.py [hunk 5]
```diff
@@ -242,14 +256,39 @@ def _maybe_warn_inputs_struct_mismatch(self, inputs, raise_exception=False):
...
+                input_layer = None
+                if (
+                    hasattr(input_tensor, "_keras_history")
+                    and input_tensor._keras_history
+                ):
+                    input_layer = input_tensor._keras_history.operation
+
+                if isinstance(input_layer, InputLayer):
+                    is_optional = getattr(input_layer, "optional", False)
+                    if not is_optional:
```
Comment (hertschuh) [inline keras/src/models/functional.py]: The attribute will be there, you don't need `getattr`:
```python
if not input_layer.optional:
```

File: keras/src/models/functional.py [hunk 6]
```diff
@@ -242,14 +256,39 @@ def _maybe_warn_inputs_struct_mismatch(self, inputs, raise_exception=False):
...
+                        input_name = getattr(input_tensor, "name", "unknown")
+                        raise ValueError(
+                            (
+                                f"The input '{input_name}' is not optional, "
+                                "but None was passed. "
+                                "Please provide a valid tensor."
+                            )
+                        )
```
Comment (hertschuh) [inline keras/src/models/functional.py]: Same thing, the `name` attribute will be there, no need for `getattr`.

File: keras/src/models/functional.py [hunk 7]
```diff
@@ -242,14 +256,39 @@ def _maybe_warn_inputs_struct_mismatch(self, inputs, raise_exception=False):
...
                 converted.append(x)
             else:
+                # Preserve dtype and sparse settings from input_tensor
+                dtype = getattr(input_tensor, "dtype", None)
+                sparse = getattr(input_tensor, "sparse", False)
                 converted.append(
-                    ops.convert_to_tensor(
-                        x, dtype=input.dtype, sparse=input.sparse
-                    )
+                    ops.convert_to_tensor(x, dtype=dtype, sparse=sparse)
```
Comment (hertschuh) [inline keras/src/models/functional.py]: I don't think these changes are needed. `dtype` and `sparse` are always there.

**Review by hertschuh — COMMENTED** [review]
Thank you for fixing this!

**Comment by MalyalaKarthik66** [comment]
@hertschuh Thanks for the feedback — I've addressed the review comments and pushed the updates.
